python/dgl/contrib/cugraph/nn/conv/relgraphconv.py

```diff
@@ -0,0 +1,383 @@
+            )
+        self.regularizer = regularizer
+        self.reset_parameters()
+
+        # others
+        self.bias = bias
+        self.activation = activation
+        self.self_loop = self_loop
+        self.layer_norm = layer_norm
+
+        # bias
+        if self.bias:
+            self.h_bias = nn.Parameter(th.Tensor(out_feat).cuda())
+            nn.init.zeros_(self.h_bias)
+
+        # layer norm
+        if self.layer_norm:
+            self.layer_norm_weight = nn.LayerNorm(
+                out_feat, elementwise_affine=True
+            )
+
+        # weight for self_loop
+        if self.self_loop:
+            self.loop_weight = nn.Parameter(
+                th.Tensor(in_feat, out_feat).cuda()
+            )
+            nn.init.xavier_uniform_(
+                self.loop_weight, gain=nn.init.calculate_gain("relu")
+            )
+
+        self.dropout = nn.Dropout(dropout)
+
+    def reset_parameters(self):
+        with th.no_grad():
+            if self.regularizer is None:
+                nn.init.uniform_(
+                    self.W,
+                    -1 / math.sqrt(self.in_feat),
+                    1 / math.sqrt(self.in_feat),
+                )
+            elif self.regularizer == "basis":
+                nn.init.uniform_(
+                    self.W,
+                    -1 / math.sqrt(self.in_feat),
+                    1 / math.sqrt(self.in_feat),
+                )
+                nn.init.xavier_uniform_(
+                    self.coeff, gain=nn.init.calculate_gain("relu")
+                )
+            else:
+                raise ValueError(
+                    f"Supported regularizer options: 'basis', "
+                    f"but got {self.regularizer}"
+                )
+
+    def forward(self, g, feat, etypes, norm=None, *, presorted=False):
+        """Forward computation.
+
+        Parameters
+        ----------
+        g : DGLGraph
+            The graph.
+        feat : torch.Tensor
+            A 2D tensor of node features. Shape: :math:`(|V|, D_{in})`.
+        etypes : torch.Tensor or list[int]
+            An 1D integer tensor of edge types. Shape: :math:`(|E|,)`.
+        norm : torch.Tensor, optional
+            An 1D tensor of edge norm value.  Shape: :math:`(|E|,)`.
+        presorted : bool, optional
+            Whether the edges of the input graph have been sorted by their
+            types. Forward on pre-sorted graph may be faster. Graphs created
+            by :func:`~dgl.to_homogeneous` automatically satisfy the condition.
+            Also see :func:`~dgl.reorder_graph` for sorting edges manually.
+
+        Returns
+        -------
+        torch.Tensor
+            New node features. Shape: :math:`(|V|, D_{out})`.
+        """
+        _sm_size = (
+            self.fanout * 2 + 3 + self.num_rels * self.num_bases
+            if self.regularizer == "basis"
+            else 0
+        )
+        _sm_size *= 8 if g.idtype == th.int64 else 4
```

Comment:
Can we have a version when the shared memory is not sufficient? Also is there any other elegant way to get the shared memory size instead of specifying it as a constant?

Reply:
> Can we have a version when the shared memory is not sufficient?

The current aggregation functions in `cugraph-ops` were designed for sampled graph, and 48KB shared memory should be enough for typical `fanout` and `num_rels` values. We are working on a full-graph variant, which can serve as a fallback option in the future.

> Also is there any other elegant way to get the shared memory size instead of specifying it as a constant?

Do you mean whether there is a better way of setting `CUDA_SM_PER_BLOCK = 49152` or a better way of calculating `_sm_size` in the `forward` function?

Reply:
Yep. I mean how to set `CUDA_SM_PER_BLOCK` because different GPUs should have different sizes.

Reply:
You are right. GPUs with different compute capabilities may have a max amount of shared memory per block ranging from 48KB to >200KB, as [Table 15](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#features-and-technical-specifications) shows. 
However, shared memory allocations over 48 KB per block are architecture-specfic and requires extra configurations to allocate. `cugraph-ops` as of now only uses up to 48 KB, so the constant here is valid, but as you said, not elegant. I can discuss with other devs to see how we could get device property and perform the check at the C layer with minimum overhead, so the Python wrapper looks less clunky. This improvement can happen in a separate PR.